resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtils.scala

```diff
@@ -0,0 +1,128 @@
+
+  val configMapNameExecutor: String = configMapName(s"spark-exec-${KubernetesUtils.uniqueID()}")
+
+  val configMapNameDriver: String = configMapName(s"spark-drv-${KubernetesUtils.uniqueID()}")
+
+  private def buildStringFromPropertiesMap(configMapName: String,
+      propertiesMap: Map[String, String]): String = {
+    val properties = new Properties()
+    propertiesMap.foreach { case (k, v) =>
+      properties.setProperty(k, v)
+    }
+    val propertiesWriter = new StringWriter()
+    properties.store(propertiesWriter,
+      s"Java properties built from Kubernetes config map with name: $configMapName")
+    propertiesWriter.toString
+  }
+
+  /**
+   * Build, file -> 'file's content' map of all the selected files in SPARK_CONF_DIR.
+   */
+  def buildSparkConfDirFilesMap(configMapName: String,
+      sparkConf: SparkConf, resolvedPropertiesMap: Map[String, String]): Map[String, String] = {
+    val loadedConfFilesMap = KubernetesClientUtils.loadSparkConfDirFiles(sparkConf)
+    // Add resolved spark conf to the loaded configuration files map.
+    if (resolvedPropertiesMap.nonEmpty) {
+      val resolvedProperties: String = KubernetesClientUtils
+        .buildStringFromPropertiesMap(configMapName, resolvedPropertiesMap)
+      loadedConfFilesMap ++ Map(Constants.SPARK_CONF_FILE_NAME -> resolvedProperties)
+    } else {
+      loadedConfFilesMap
+    }
+  }
+
+  def buildKeyToPathObjects(confFilesMap: Map[String, String]): Seq[KeyToPath] = {
+    confFilesMap.map {
+      case (fileName: String, _: String) =>
+        val filePermissionMode = 420  // 420 is decimal for octal literal 0644.
+        new KeyToPath(fileName, filePermissionMode, fileName)
+    }.toList.sortBy(x => x.getKey) // List is sorted to make mocking based tests work
+  }
+
+  /**
+   * Build a Config Map that will hold the content for environment variable SPARK_CONF_DIR
+   * on remote pods.
+   */
+  def buildConfigMap(configMapName: String, confFileMap: Map[String, String],
+      withLabels: Map[String, String] = Map()): ConfigMap = {
+    new ConfigMapBuilder()
+      .withNewMetadata()
+        .withName(configMapName)
+        .withLabels(withLabels.asJava)
+        .endMetadata()
+      .addToData(confFileMap.asJava)
+      .build()
+  }
+
+  private def loadSparkConfDirFiles(conf: SparkConf): Map[String, String] = {
+    val confDir = Option(conf.getenv(ENV_SPARK_CONF_DIR)).orElse(
+      conf.getOption("spark.home").map(dir => s"$dir/conf"))
+    if (confDir.isDefined) {
+      val confFiles = listConfFiles(confDir.get)
+      logInfo(s"Spark configuration files loaded from $confDir : ${confFiles.mkString(",")}")
+      confFiles.map { file =>
+        val source = Source.fromFile(file)(Codec.UTF8)
+        val mapping = (file.getName -> source.mkString)
+        source.close()
+        mapping
+      }.toMap
+    } else {
+      logInfo(s"Spark configuration directory is not detected, please set env:$ENV_SPARK_CONF_DIR")
```

Comment:
For the existing users, I guess we don't need to give this additional direction because they don't need this.